src/evrard/evrard.cpp

```diff
@@ -101,126 +105,120 @@ int main(int argc, char** argv)
-
 #ifdef SPH_EXA_USE_ASCENT
     AscentAdaptor::Initialize(d, domain.startIndex());
     std::cout << "AscentInitialize\n";
 #endif
+
+    const size_t nTasks = 64;
+    const size_t ngmax  = 150;
+    const size_t ng0    = 100;
+
+    Propagator propagator(nTasks, ngmax, ng0, domain.nParticles(), output, d.rank);
+
     if (d.rank == 0) std::cout << "Starting main loop." << std::endl;
 
     totalTimer.start();
     for (d.iteration = 0; d.iteration <= maxStep; d.iteration++)
     {
-        timer.start();
-        domain.sync(d.codes, d.x, d.y, d.z, d.h, d.m, d.mui, d.u, d.vx, d.vy, d.vz, d.x_m1, d.y_m1, d.z_m1, d.du_m1,
-                    d.dt_m1);
-        timer.step("domain::sync");
-
-        d.resize(domain.nParticlesWithHalos()); // also resize arrays not listed in sync, even though space for halos is
-                                                // not needed
-        // domain.exchangeHalos(d.m);
-        std::fill(begin(d.m), begin(d.m) + domain.startIndex(), d.m[domain.startIndex()]);
-        std::fill(begin(d.m) + domain.endIndex(), begin(d.m) + domain.nParticlesWithHalos(), d.m[domain.startIndex()]);
-
-        taskList.update(domain.startIndex(), domain.endIndex());
-        timer.step("updateTasks");
-        findNeighborsSfc(taskList.tasks, d.x, d.y, d.z, d.h, d.codes, domain.box());
-        timer.step("FindNeighbors");
-        computeDensity<Real>(taskList.tasks, d, domain.box());
-        timer.step("Density");
-        computeEquationOfStateEvrard<Real>(taskList.tasks, d);
-        timer.step("EquationOfState");
-        domain.exchangeHalos(d.vx, d.vy, d.vz, d.ro, d.p, d.c);
-        timer.step("mpi::synchronizeHalos");
-        computeIAD<Real>(taskList.tasks, d, domain.box());
-        timer.step("IAD");
-        domain.exchangeHalos(d.c11, d.c12, d.c13, d.c22, d.c23, d.c33);
-        timer.step("mpi::synchronizeHalos");
-        computeMomentumAndEnergyIAD<Real>(taskList.tasks, d, domain.box());
-        timer.step("MomentumEnergyIAD");
-        d.egrav = domain.addGravityAcceleration(d.x, d.y, d.z, d.h, d.m, d.g, d.grad_P_x, d.grad_P_y, d.grad_P_z);
-        // temporary sign fix, see note in ParticlesData
-        d.egrav = (d.g > 0.0) ? d.egrav : -d.egrav;
-        timer.step("Gravity");
-        computeTimestep<Real, TimestepPress2ndOrder<Real, Dataset>>(taskList.tasks, d);
-        timer.step("Timestep"); // AllReduce(min:dt)
-        computePositions<Real, computeAcceleration<Real, Dataset>>(taskList.tasks, d, domain.box());
-        timer.step("UpdateQuantities");
-        computeTotalEnergy<Real>(taskList.tasks, d);
-        d.etot += d.egrav;
-        timer.step("EnergyConservation"); // AllReduce(sum:ecin,ein)
-        updateSmoothingLength<Real>(taskList.tasks, d);
-        timer.step("UpdateSmoothingLength");
-
-        size_t totalNeighbors = neighborsSum(taskList.tasks);
+        propagator.hydroStepGravity<DomainType, Dataset, Real>(domain, d);
```

Comment:
```suggestion
        propagator.hydroStepGravity(domain, d);
```
Should be possible once the last template parameter is removed